### ISSUE TITLE
Add back io::copy file to pipe optimization 

### ISSUE THREAD

**Issue by NobodyXu** [rust-lang/rust#128300]
So to sum up, I think we can add back optimization via:
 - if the file is on squashfs/erofs, then it's immutable (detectable via [`statfs`](https://man7.org/linux/man-pages/man2/statfs.2.html)
 - if the is sealed with `F_SEAL_WRITE `(prevent modifying of existing contents, shrinking and appending is allowed) and `F_SEAL_SEAL` (not changing of seals), detectable via `fcntl(fd, F_GET_SEALS)`
 - [if it's on read-only btrfs subvolume (e.g. snapshot)](https://unix.stackexchange.com/a/375646), likewise we can do the same for zfs, nilfs, bcachefs
 - If the file is `/dev/zero`, or from other pseudo fs (e.g. procfs, sysfe), then I think we can assume it's immutable

Otherwise, we'd fallback to:
 - read + write if the file is really small
 - splice to an intermediate pipe with F_MOVE and then splice to the pipe without F_MOVE if the file is more than 1-2 pages
 - if the fs supports reflink and file is really large, then we could create a O_TMPFILE on the system, copy_file_range to it and then splice it

Also, if the file is a named fifo, then just a splice without F_MOVE would work fine.

_Originally posted by @NobodyXu in https://github.com/rust-lang/rust/issues/108283#issuecomment-2254439139_

**Comment by the8472** [comment]
> splice to an intermediate pipe with F_MOVE and then splice to the pipe without F_MOVE if

We never used F_MOVE and it was still a problem. So I'm skeptical it'll work at all but maybe it depends on the direction (from/to pipe).

> then it's immutable (detectable via [statfs](https://man7.org/linux/man-pages/man2/statfs.2.html)
> detectable via fcntl(fd, F_GET_SEALS)

Those cost extra syscalls, which could be unnecessary overhead for small files. So it probably only makes sense to even try these things for large files.

And many of those cases seem rare that it's not worth it. Like who serves e.g. a webserver from a btrfs snapshot?

**Comment by NobodyXu** [comment]
Yeah I think the most common case is a sealed file, I was probably listing too many cases during my brain storming 😂

> We never used F_MOVE and it was still a problem. So I'm skeptical it'll work at all but maybe it depends on the direction (from/to pipe).

Hmm it seems that `SPLICE_F_MOVE` is a no-op according to [man page](https://man7.org/linux/man-pages/man2/splice.2.html), so I think using an intermediate pipe to workaround this will not work.

So the only possible workarond **for a really large file and fs supporting reflink** would be to use `copy_file_range` to a `O_TMPFILE` and then `splice`.

> So it probably only makes sense to even try these things for large files.

Agree, if it is small enough, `read` + `write` is OK.

**Comment by NobodyXu** [comment]
I think we might want someway for the user to tell `std::io::copy` that they know that the file won't change before they sent out the data, the application itself could make some reasonable assumption about this

**Comment by the8472** [comment]
> Currently, file seals can be applied only to a file descriptor returned by [memfd_create(2)](https://man7.org/linux/man-pages/man2/memfd_create.2.html) 

Also doesn't seem like a common thing.

**Comment by NobodyXu** [comment]
> > Currently, file seals can be applied only to a file descriptor returned by [memfd_create(2)](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
> 
> Also doesn't seem like a common thing.

Thanks I missed that one, if so then file seals seems very uncommon

**Comment by NobodyXu** [comment]
> And the files might not even be marked read-only there.

Agree, the more I look at it, the more I think we need another API like rust-lang/libs-team#202 propose, which will always assume using `splice` is ok:

```rust
pub struct ImmutableFile(File);
```

That's the simplest way I can think of, no need for another `copy` function as we just want to promise that the file is indeed immutable.

**Comment by the8472** [comment]
I think we don't want to promise anything that relies on specialization until there's a plan to stabilize specialization or at least parts thereof.

**Comment by NobodyXu** [comment]
> I think we don't want to promise anything that relies on specialization until there's a plan to stabilize specialization or at least parts thereof.

That makes sense, in that case maybe a dedicated API on `File`?

```rust
impl File {
    /// This is similar to [`std::io::copy`], except it will attempt to use `splice` on Linux or `sendfile` on Unix
    /// to avoid copying data if possible.
    ///
    /// NOTE that if you change the content of the file after this call, it might also change the data in `writer`.
    pub fn copy(&self, writer: &mut W) -> io::Result<u64>
    where
        W: Write + ?Sized;
}
```

**Comment by the8472** [comment]
That would still require specialization for the writer side. The only way to do this with stable rust is to require both sides to implement all the necessary traits which expose the internal buffers and so on. What `CopyRead` and `CopyWrite` currently do. And that's some gnarly machinery that might not be the right fit for std, considering that only useful on some platforms.
Nobody has even generalized the current optimizations to other platforms.

**Comment by NobodyXu** [comment]
> That would still require specialization for the writer side.

Yeah but isn't that same as `std::io::copy` where we specialise internally and not exposing it?

**Comment by the8472** [comment]
We don't guarantee that that happens though, so we can remove it if specialization becomes an issue.

**Comment by NobodyXu** [comment]
> We don't guarantee that that happens though, so we can remove it if specialization becomes an issue.

So...maybe we just add it, saying that we assume the `File` is immutable during the `File::copy`, and leave out all the details?

If specialisation is removed, it'd be just a convenient method of `std::io::copy`

**Comment by the8472** [comment]
Except that we'd have to make that distinction very clear to avoid people getting corrupt data. And adding warnings and choosing a different name for something we can't promise probably isn't worth it.
Now that I have written that I should probably close that ACP....

**Comment by oech3** [comment]
Does IORING_OP_SPLICE have interface to know the time when content reached to the target?

**Comment by oech3** [comment]
I think `std::io::sink` can be optimized by splice to `/dev/null` since it does not require to check content at least for.
(uutils `wc -c` uses splice for byte count...)

**Comment by oech3** [comment]
Is the ordering issue solved just by writing last 1 byte by read/write?

**Comment by the8472** [comment]
The problem is unsent/unacknowledged bytes still sitting in a socket send buffer when the call returns, which can be more than 1 byte. So we'd have to ensure that enough bytes get written to displace the whole buffer.

And if the pipeline looks like `file -(io::copy)-> pipe -(something else)-> socket` then we wouldn't even be able to look at what's going on in the socket. So it's just hard to figure out at which point it's safe.

Some related discussion that happened to come up recently: https://lore.kernel.org/linux-fsdevel/0c9f58f1-52f6-481c-829f-6f5a70ffff7f@infinite-source.de/

**Comment by oech3** [comment]
I guess we need the syscall to get reflink count of pages (incremented by tee(2)).
Just waiting last page's reflink count being 1.

**Comment by oech3** [comment]
So is input -> pipe (as io::copy's buffer) -> non-stream by io::copy OK as io::copy can wait until pipe becomes empty?
